[iOS] Address fullscreen warning UI feedback

jernoble opened the pull request:
#### ad9eeb0d8d0e77fea0edd9be5888b24b083465bd
<pre>
[iOS] Address fullscreen warning UI feedback
<a href="https://bugs.webkit.org/show_bug.cgi?id=269340">https://bugs.webkit.org/show_bug.cgi?id=269340</a>
<a href="https://rdar.apple.com/117099085">rdar://117099085</a>

Reviewed by Aditya Keerthi.

Address feedback re: the fullscreen warning panel is visible upon entering fullscreen
by making two changes to the warning panel:

1. Move the panel to the top third of the screen, to avoid both media controls in the
   center of the screen, as well as controls along the bottom.

2. Allow the user to dismiss the warning panel by tapping on it; but ensure the panel
   is visible for a minimum amount of time, even if tapped.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController invalidate]):
(-[WKFullScreenViewController showBanner]):
(-[WKFullScreenViewController hideBanner]):
(-[WKFullScreenViewController _minimumHideDelayReached]):
(-[WKFullScreenViewController _bannerDismissalRecognized:]):
(-[WKFullScreenViewController loadView]):

Canonical link: <a href="https://commits.webkit.org/274684@main">https://commits.webkit.org/274684@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d8218ff2a123b5a4ea42568c03c02c0c61e09c4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39645 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18624 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42001 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42180 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35545 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21523 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15953 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33100 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40219 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15715 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34313 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13627 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13613 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43457 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36005 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35600 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39394 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14457 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11911 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37663 "Found 1 new API test failure: /TestWTF:WTF_ParkingLot.UnparkOneOneFast (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16063 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8902 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16112 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15720 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -581,6 +612,11 @@ - (void)loadView
 #endif
     _banner = WTFMove(banner);
+
+    _bannerTapToDismissRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(_bannerDismissalRecognized:)];
```
`adoptNS`

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -130,6 +131,9 @@ @implementation WKFullScreenViewController {
     RetainPtr<_WKInsetLabel> _bannerLabel;
+    RetainPtr<UITapGestureRecognizer> _bannerTapToDismissRecognizer;
+    BOOL _blockHideBanner;
+    BOOL _hideBannerAfterMinimumHideDelayReached;
```
I feel the logic would be easier to follow if we simply stored a timestamp of last banner presentation.

If the difference is less than `minimumHideDelay`, you would call `performSelector:afterDelay` with the difference.

jernoble replied:
TBH, I was torn between the approach in the patch and this one you mention here. Let me try out the calculation approach.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -321,6 +331,27 @@ - (void)hideBanner
+        [self hideBanner];
+        _hideBannerAfterMinimumHideDelayReached = NO;
+    }
+}
```
I think this also needs to be inside a `#if ENABLE(FULLSCREEN_DISMISSAL_GESTURES)`?

jernoble replied:
Unnecessary after the new time-delta approach.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -321,6 +331,27 @@ - (void)hideBanner
+    }
+}
+
+- (void)_bannerDismissalRecognized:(NSNotification*)notification
```
Can't the method declaration also be behind `#if ENABLE(FULLSCREEN_DISMISSAL_GESTURES)`?

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
```diff
@@ -321,6 +327,19 @@ - (void)hideBanner
+        return;
+    }
+
+    [self performSelector:@selector(hideBanner) withObject:nil afterDelay:remainingDelay.seconds()];
```
Do we also need a `cancelPreviousPerformRequestsWithTarget` when the banner is displayed, so that we don't run into a scenario where we dismiss immediately when the banner is brought up again?

jernoble replied:
Both `-hideBanner` and `-showBanner` will `cancelPrevious...`. So I don't think there's a scenario where the tap-to-dismiss behavior will cause the banner to immediately hide after showing.

pxlcoder replied:
Thanks!

jernoble committed:
[iOS] Address fullscreen warning UI feedback
https://bugs.webkit.org/show_bug.cgi?id=269340
rdar://117099085

Reviewed by Aditya Keerthi.

Address feedback re: the fullscreen warning panel is visible upon entering fullscreen
by making two changes to the warning panel:

1. Move the panel to the top third of the screen, to avoid both media controls in the
   center of the screen, as well as controls along the bottom.

2. Allow the user to dismiss the warning panel by tapping on it; but ensure the panel
   is visible for a minimum amount of time, even if tapped.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController invalidate]):
(-[WKFullScreenViewController showBanner]):
(-[WKFullScreenViewController hideBanner]):
(-[WKFullScreenViewController _minimumHideDelayReached]):
(-[WKFullScreenViewController _bannerDismissalRecognized:]):
(-[WKFullScreenViewController loadView]):

Canonical link: https://commits.webkit.org/274684@main
